### PR TITLE
gRPC changes for FSMaster services

### PR DESCRIPTION
**Pull request by ggezer**

### CODE REVIEW

File: core/transport-options/src/grpc/file_system_master_options.proto [hunk 1]
```diff
@@ -0,0 +1,172 @@
+syntax = "proto2";
```
Comment (madanadit) [inline core/transport-options/src/grpc/file_system_master_options.proto]: we should move to proto3 soon

**Comment by ggezer** [comment]
@madanadit, @aaudiber PTAL.

**High level summary of changes:**
- Remove option wrappers on client side.
- Replace option wrappers on master side with OperationContext implementations.
- Implement merging of master options with given options. So that going forward clients don't need to send defaults. (Client side is not modified yet to not send defaults)

File: core/client/fs/src/main/java/alluxio/client/ReadType.java [hunk 2]
```diff
@@ -82,4 +84,15 @@ public boolean isCache() {
   public boolean isPromote() {
     return mValue == CACHE_PROMOTE.mValue;
   }
+
+  // TODO(ggezer) Eliminate below parsers/converters for ReadType
+  private static final String sProtoNameHeader = "READ_";
```
Comment (aaudiber) [inline core/client/fs/src/main/java/alluxio/client/ReadType.java]: why is this needed?

File: core/client/fs/src/main/java/alluxio/client/file/FileSystemClientOptions.java [hunk 3]
```diff
@@ -0,0 +1,234 @@
...
+import alluxio.wire.LoadMetadataType;
+import alluxio.wire.TtlAction;
+
+import javax.annotation.concurrent.ThreadSafe;
+
+/**
+ * The file system class to set default options for client.
+ */
+@ThreadSafe
+public final class FileSystemClientOptions {
```
Comment (aaudiber) [inline core/client/fs/src/main/java/alluxio/client/file/FileSystemClientOptions.java]: Is this temporary? I thought we were going to fill in defaults on the server side

File: core/server/master/src/main/java/alluxio/master/file/options/CompleteFileContext.java [hunk 4]
```diff
@@ -0,0 +1,108 @@
...
+
+/**
+ * Wrapper for {@link CompleteFilePOptions} with additional context data.
+ */
+public class CompleteFileContext extends OperationContext<CompleteFilePOptions.Builder> {
+
+  private long mOperationTimeMs;
+  private UfsStatus mUfsStatus;
+
+  // Prevent instantiation
```
Comment (aaudiber) [inline core/server/master/src/main/java/alluxio/master/file/options/CompleteFileContext.java]: this is already accomplished by the other private constructor

File: core/server/master/src/main/java/alluxio/master/file/options/CompleteFileContext.java [hunk 5]
```diff
@@ -0,0 +1,108 @@
...
+  }
+
+  /**
+   * Merges and embeds the given {@link CompleteFilePOptions} with the corresponding master options.
+   *
+   * @param optionsBuilder Builder for proto {@link CompleteFilePOptions} to embed
+   * @return the instance of {@link CompleteFileContext} with default values for master
+   */
+  public static CompleteFileContext defaults(CompleteFilePOptions.Builder optionsBuilder) {
+    CompleteFilePOptions masterOptions = FileSystemMasterOptions.getCompleteFileOptions();
```
Comment (aaudiber) [inline core/server/master/src/main/java/alluxio/master/file/options/CompleteFileContext.java]: naming is confusing here - `FileSystemMasterOptions.getCompleteFileOptions()` gets default option values but doesn't mention anything about defaults, while `CompleteFileContext.defaults(builder)` is named `defaults`, but returns non-default values

One idea:
`defaults` -> `applyDefaults`
`FileSystemMasterOptions.getCompleteFileOptions` -> `FileSystemMasterOptions.completeFileDefaults()`

File: core/server/master/src/main/java/alluxio/master/file/DefaultFileSystemMaster.java [hunk 6]
```diff
@@ -803,17 +804,18 @@ public FileInfo getFileInfo(AlluxioURI path, GetStatusOptions options)
...
-        checkLoadMetadataOptions(options.getLoadMetadataType(), inodePath.getUri());
+        checkLoadMetadataOptions(context.getOptions().getLoadMetadataType(), inodePath.getUri());
         loadMetadataIfNotExist(rpcContext, inodePath,
-            MASTER_OPTIONS.getLoadMetadataOptions().setCreateAncestors(true).setCommonOptions(
-                MASTER_OPTIONS.getCommonOptions().setTtl(options.getCommonOptions().getTtl())
-                    .setTtlAction(options.getCommonOptions().getTtlAction())));
+            LoadMetadataContext.defaults(LoadMetadataPOptions.newBuilder().setCreateAncestors(true)
+                .setCommonOptions(FileSystemMasterCommonPOptions.newBuilder()
+                    .setTtl(context.getOptions().getCommonOptions().getTtl())
+                    .setTtlAction(context.getOptions().getCommonOptions().getTtlAction()))));
```
Comment (aaudiber) [inline core/server/master/src/main/java/alluxio/master/file/DefaultFileSystemMaster.java]: probably worth saving `context.getOptions()` and `context.getOptions.getCommonOptions()` to local variables since they are used many times.

File: core/server/master/src/main/java/alluxio/master/file/options/CompleteFileContext.java (hunk 5 above)
Comment (aaudiber) [reply core/server/master/src/main/java/alluxio/master/file/options/CompleteFileContext.java]: we could also move the `FileSystemMasterOptions.getCompleteFileOptions` to be `private static CompleteFileContext.defaults()`. I don't think it needs to be exposed outside of `CompleteFileContext`

File: bin/alluxio [hunk 7]
```diff
@@ -398,9 +398,13 @@ function main {
...
+    rm -rf ${BIN}/../core/transport-options/src/main/java/alluxio/grpc
+    for src_file in $(find ${BIN}/../core/transport-options/src/grpc -name '*.proto' -type f); do
+      protoc --java_out=${BIN}/../core/transport-options/src/main/java --proto_path=${BIN}/../core/transport-options/src/grpc --plugin=protoc-gen-grpc-java=${BIN}/protoc-gen-grpc-java-1.10.1-osx-x86_64.exe --grpc-java_out=${BIN}/../core/transport-options/src/main/java ${src_file}
+    done
     rm -rf ${BIN}/../core/transport/src/main/java/alluxio/grpc
     for src_file in $(find ${BIN}/../core/transport/src/grpc -name '*.proto' -type f); do
-      protoc --java_out=${BIN}/../core/transport/src/main/java --proto_path=${BIN}/../core/transport/src/grpc --plugin=protoc-gen-grpc-java=${BIN}/protoc-gen-grpc-java-1.10.1-osx-x86_64.exe --grpc-java_out=${BIN}/../core/transport/src/main/java ${src_file}
+      protoc --java_out=${BIN}/../core/transport/src/main/java --proto_path=${BIN}/../core/transport/src/grpc --proto_path=${BIN}/../core/transport-options/src/grpc --plugin=protoc-gen-grpc-java=${BIN}/protoc-gen-grpc-java-1.10.1-osx-x86_64.exe --grpc-java_out=${BIN}/../core/transport/src/main/java ${src_file}
```
Comment (madanadit) [inline bin/alluxio]: is there a better way than referencing `${BIN}/protoc-gen-grpc-java-1.10.1-osx-x86_64.exe`?

File: core/client/fs/src/main/java/alluxio/client/AlluxioStorageType.java [hunk 8]
```diff
@@ -11,6 +11,9 @@
 
 package alluxio.client;
 
+import alluxio.grpc.WritePType;
```
Comment (madanadit) [inline core/client/fs/src/main/java/alluxio/client/AlluxioStorageType.java]: both these are not used?

File: core/client/fs/src/main/java/alluxio/client/UnderStorageType.java [hunk 9]
```diff
@@ -11,6 +11,9 @@
 
 package alluxio.client;
 
+import alluxio.grpc.WritePType;
```
Comment (madanadit) [inline core/client/fs/src/main/java/alluxio/client/UnderStorageType.java]: same here

File: core/server/master/src/main/java/alluxio/master/file/options/CompleteFileContext.java (hunk 5 above)
Comment (ggezer) [reply core/server/master/src/main/java/alluxio/master/file/options/CompleteFileContext.java]: Ultimate plan is to get rid of FileSystemMasterOptions and move the definitions to corresponding operation contexts. I'll do that all at once for all operation types.

File: core/client/fs/src/main/java/alluxio/client/file/FileSystemClientOptions.java (hunk 3 above)
Comment (ggezer) [reply core/client/fs/src/main/java/alluxio/client/file/FileSystemClientOptions.java]: Yes, but depends.
It's temporarily used here to provide client defaults. (Only the infra is done but the clients are still supplying defaults)

However, if there are discrepancies between client/master default values, then we would need to use this to merge client provided settings with client defaults, which would ultimately be merged with master defaults.

I'll tackle getting rid of client defaults as a separate work.

File: core/client/fs/src/main/java/alluxio/client/ReadType.java (hunk 2 above)
Comment (ggezer) [reply core/client/fs/src/main/java/alluxio/client/ReadType.java]: gRPC enums are defined in a global scope. Defining them without some kind of header invites conflicts for commonly used enum values. (E.g NONE,ALL, etc...)

Comment (ggezer) [reply core/client/fs/src/main/java/alluxio/client/ReadType.java]: PS: I don't intend to keep this header trick for longer. Once we start eliminating wire types, they should go away naturally.

File: core/server/master/src/main/java/alluxio/master/file/options/CompleteFileContext.java (hunk 5 above)
Comment (ggezer) [reply core/server/master/src/main/java/alluxio/master/file/options/CompleteFileContext.java]: After reconsideration I feel like *Context.defaults() overload for taking proto options makes sense. Parameter javadoc also clearly states that given options are due for merging with defaults.

Comment (aaudiber) [reply core/server/master/src/main/java/alluxio/master/file/options/CompleteFileContext.java]: I still think it's misleading. We have historically used `.defaults()` methods on all of our option classes to return the default options. How about `CompleteFileContext.create(completeFileOptions)`?

File: bin/alluxio (hunk 7 above)
Comment (ggezer) [reply bin/alluxio]: Making maven to generate the files maybe?

File: core/server/master/src/main/java/alluxio/master/file/options/CompleteFileContext.java [hunk 10]
```diff
@@ -0,0 +1,104 @@
...
+    mUfsStatus = null;
+  }
+
+  /**
+   * Merges and embeds the given {@link CompleteFilePOptions} with the corresponding master options.
+   *
+   * @param optionsBuilder Builder for proto {@link CompleteFilePOptions} to merge with defaults
+   * @return the instance of {@link CompleteFileContext} with default values for master
+   */
+  public static CompleteFileContext defaults(CompleteFilePOptions.Builder optionsBuilder) {
```
Comment (aaudiber) [inline core/server/master/src/main/java/alluxio/master/file/options/CompleteFileContext.java]: Can we call this `create`, `applyDefaults`, or `createWithDefaults`? We typically use verbs for methods that process an input.

Comment (ggezer) [reply core/server/master/src/main/java/alluxio/master/file/options/CompleteFileContext.java]: Agreed to naming.

Considering the overall comment you gave, on keeping the given defaults as is, how about keeping the proto options as builder but provide another method from base OperationContext to expose given defaults. So in the end we would have above three methods for creating the contexts and each context would have getOriginalOptions() method that returns whatever was provided by the caller before merging or modification. 

Using a builder makes setting proto fields very easy. It's not just to avoid extra .build() call at the end but to be able to set proto fields line-by-line as below, which is very usefull in places where there are many proto fields (e.g. InodeTree).

context = SomeContext.createWithDefaults();
context.getOptions().setQwe();
context.getOptions().setAsd();

compared to;

context = SomeContext.createWithDefaults();
context.setOptions(context.getOptions().toBuilder().setQwe().setAsd().build);

Comment (aaudiber) [reply core/server/master/src/main/java/alluxio/master/file/options/CompleteFileContext.java]: At the start of the RPC we receive options from the client and merge them with the defaults to create a complete `*POptions` object. Most RPCs never modify this `*POptions` object. Ideally no RPCs would modify it. We can guarantee this by making it immutable, simplifying our lives greatly because we don't need to worry about all the options changing.

Modifying options during an RPC call is rare. I only see three places where we need to do it:

https://github.com/Alluxio/alluxio/pull/8122/files#diff-7f82122f923de058e3c4a20b43cd7e4eR807
https://github.com/Alluxio/alluxio/pull/8122/files#diff-7f82122f923de058e3c4a20b43cd7e4eR890
https://github.com/Alluxio/alluxio/pull/8122/files#diff-7f82122f923de058e3c4a20b43cd7e4eR921

All 3 of these are related to loadMetadataType/loadDescendantType, which are effectively deprecated now that we have ufs sync. We shouldn't base the convention off of these outliers. Immutability is valuable for preventing bugs, and we have an opportunity here to make dozens of fields (every option) immutable at a low cost code-wise

File: bin/alluxio (hunk 7 above)
Comment (madanadit) [reply bin/alluxio]: i would prefer pre-generated files. we can defer this to following PRs

File: core/server/master/src/main/java/alluxio/master/file/DefaultFileSystemMaster.java [hunk 11]
```diff
@@ -2451,37 +2454,42 @@ private void loadFileMetadataInternal(RpcContext rpcContext, LockedInodePath ino
...
-        .setRecursive(options.isCreateAncestors())
-        .setMetadataLoad(true)
-        .setPersisted(true)
-        .setTtl(options.getCommonOptions().getTtl())
-        .setTtlAction(options.getCommonOptions().getTtlAction())
-        .setOwner(ufsStatus.getOwner())
-        .setGroup(ufsStatus.getGroup());
-    short ufsMode = ufsStatus.getMode();
+    CreateFileContext createFileContext = CreateFileContext.defaults();
+    createFileContext.getOptions().setBlockSizeBytes(ufsBlockSizeByte);
```
Comment (madanadit) [inline core/server/master/src/main/java/alluxio/master/file/DefaultFileSystemMaster.java]: why set createFileContext instead of the way it was?

File: core/server/master/src/main/java/alluxio/master/file/DefaultFileSystemMaster.java [hunk 12]
```diff
@@ -2451,37 +2454,42 @@ private void loadFileMetadataInternal(RpcContext rpcContext, LockedInodePath ino
...
-      CompleteFileOptions completeOptions =
-          MASTER_OPTIONS.getCompleteFileOptions().setUfsLength(ufsLength).setUfsStatus(ufsStatus);
+      createFileInternal(rpcContext, inodePath, createFileContext);
+
+      CompleteFileContext completeCtx =
+          CompleteFileContext.defaults(CompleteFilePOptions.newBuilder().setUfsLength(ufsLength))
+              .setUfsStatus(context.getUfsStatus());
       if (ufsLastModified != null) {
-        completeOptions.setOperationTimeMs(ufsLastModified);
+        completeCtx.setOperationTimeMs(ufsLastModified);
```
Comment (madanadit) [inline core/server/master/src/main/java/alluxio/master/file/DefaultFileSystemMaster.java]: nit: consistent w/ naming => `completeContext`

File: core/server/master/src/main/java/alluxio/master/file/DefaultFileSystemMaster.java (hunk 11 above)
Comment (madanadit) [reply core/server/master/src/main/java/alluxio/master/file/DefaultFileSystemMaster.java]: i see. should we set all `options` fields in one go?

File: bin/alluxio (hunk 7 above)
Comment (ggezer) [reply bin/alluxio]: I didn't mean generating during the build but invoking the generation phase only when we want to regenerate, not triggering the generation in the regular build.

File: core/server/master/src/main/java/alluxio/master/file/FileSystemMasterClientRestServiceHandler.java [hunk 13]
```diff
@@ -14,17 +14,25 @@
 import alluxio.AlluxioURI;
 import alluxio.Constants;
 import alluxio.RestUtils;
+import alluxio.client.file.FileSystemClientOptions;
```
Comment (madanadit) [inline core/server/master/src/main/java/alluxio/master/file/FileSystemMasterClientRestServiceHandler.java]: why client options default?

File: core/server/master/src/main/java/alluxio/master/file/InodeTtlChecker.java [hunk 14]
```diff
@@ -98,9 +101,10 @@ public void heartbeat() {
                 // public delete method will lock the path, and check WRITE permission required at
                 // parent of file
                 if (inode.isDirectory()) {
-                  mFileSystemMaster.delete(path, DeleteOptions.defaults().setRecursive(true));
+                  mFileSystemMaster.delete(path,
+                      DeleteContext.defaults(DeletePOptions.newBuilder().setRecursive(true)));
                 } else {
-                  mFileSystemMaster.delete(path, DeleteOptions.defaults().setRecursive(false));
+                  mFileSystemMaster.delete(path, DeleteContext.defaults());
```
Comment (madanadit) [inline core/server/master/src/main/java/alluxio/master/file/InodeTtlChecker.java]: is this intentional?

File: core/server/master/src/main/java/alluxio/master/file/meta/InodeDirectory.java [hunk 15]
```diff
@@ -17,12 +17,13 @@
 import alluxio.collections.UniqueFieldIndex;
 import alluxio.exception.InvalidPathException;
 import alluxio.master.ProtobufUtils;
-import alluxio.file.options.CreateDirectoryOptions;
+import alluxio.master.file.options.CreateDirectoryContext;
```
Comment (madanadit) [inline core/server/master/src/main/java/alluxio/master/file/meta/InodeDirectory.java]: should we rename the package to `context`?

File: core/server/master/src/main/java/alluxio/master/file/options/CompleteFileContext.java (hunk 10 above)
Comment (ggezer) [reply core/server/master/src/main/java/alluxio/master/file/options/CompleteFileContext.java]: @madanadit, this discussion is related to your question :
> should we set all options fields in one go?

File: core/transport-options/src/grpc/file_system_master_options.proto (hunk 1 above)
Comment (ggezer) [reply core/transport-options/src/grpc/file_system_master_options.proto]: It's on my task list for the consolidation work after security.

File: core/client/fs/src/main/java/alluxio/client/file/BaseFileSystem.java [hunk 16]
```diff
@@ -145,26 +150,27 @@ public FileOutStream createFile(AlluxioURI path, CreateFileOptions options)
     } finally {
       mFileSystemContext.releaseMasterClient(masterClient);
     }
-    OutStreamOptions outStreamOptions = options.toOutStreamOptions();
-    outStreamOptions.setUfsPath(status.getUfsPath());
-    outStreamOptions.setMountId(status.getMountId());
-    outStreamOptions.setAcl(status.getAcl());
+
     try {
+      OutStreamOptions outStreamOptions = new OutStreamOptions(options);
```
[the remark this answers is not in the document]
Comment (ggezer) [reply core/client/fs/src/main/java/alluxio/client/file/BaseFileSystem.java]: Because now OutStream is created by a constructor that takes proto options during which we try to create a class instance using class name. 

I wrapped that code in OutStreamOptions with try-catch and throw RuntimeException. Code is out of the try-catch now.

File: core/client/fs/src/main/java/alluxio/client/file/options/InStreamOptions.java [hunk 17]
```diff
@@ -11,52 +11,81 @@
...
+   * Creates with given {@link OpenFilePOptions} instance.
+   * @param status URI status
+   * @param options OpenFile options
    */
-  public InStreamOptions(URIStatus status, OpenFileOptions options) {
+  public InStreamOptions(URIStatus status, OpenFilePOptions options) {
     mStatus = status;
-    mOptions = options;
+    mProtoOptions = options;
+    CreateOptions blockLocationPolicyCreateOptions = CreateOptions.defaults()
```
[the remark this answers is not in the document]
Comment (ggezer) [reply core/client/fs/src/main/java/alluxio/client/file/options/InStreamOptions.java]: JobUtils.loadBlock also calls that along with many other tests.

File: core/client/fs/src/test/java/alluxio/client/block/AlluxioBlockStoreTest.java [hunk 18]
```diff
@@ -94,11 +95,19 @@
    * A mock class used to return controlled result when selecting workers.
    */
   @ThreadSafe
-  private static class MockFileWriteLocationPolicy
+  public static class MockFileWriteLocationPolicy
```
[the remark this answers is not in the document]
Comment (ggezer) [reply core/client/fs/src/test/java/alluxio/client/block/AlluxioBlockStoreTest.java]: I thought it was necessary now that creating an instance of it was done by Class.forName within the InStreamOptions. I tried private and it worked. Reverted it.

File: core/server/master/src/main/java/alluxio/master/file/DefaultFileSystemMaster.java (hunk 11 above)
Comment (ggezer) [reply core/server/master/src/main/java/alluxio/master/file/DefaultFileSystemMaster.java]: We didn't have to. However, due to discussions for immutable of options, that's what's going to happen. So yes, we should set it all in one go.

File: core/server/common/src/main/java/alluxio/master/AbstractNonJournaledMaster.java [hunk 19]
```diff
@@ -46,6 +47,6 @@ public void resetState() {
 
   @Override
   public Iterator<JournalEntry> getJournalEntryIterator() {
-    return Iterators.emptyIterator();
+    return Iterators.unmodifiableIterator(Collections.emptyIterator());
```
[the remark this answers is not in the document]
Comment (ggezer) [reply core/server/common/src/main/java/alluxio/master/AbstractNonJournaledMaster.java]: Due to Guava version bump, emptyIterator() is no longer there. This is a functional equivalent of it.

File: bin/alluxio (hunk 7 above)
Comment (ggezer) [reply bin/alluxio]: I put this on my task list for the consolidation work.

File: core/client/fs/src/main/java/alluxio/client/file/BaseFileSystem.java [hunk 20]
```diff
@@ -145,26 +150,27 @@ public FileOutStream createFile(AlluxioURI path, CreateFileOptions options)
...
+      outStreamOptions.setUfsPath(status.getUfsPath());
+      outStreamOptions.setMountId(status.getMountId());
+      outStreamOptions.setAcl(status.getAcl());
       return new FileOutStream(path, outStreamOptions, mFileSystemContext);
     } catch (Exception e) {
       delete(path);
-      throw e;
+      throw new AlluxioException(e.getMessage(), e);
```
[the remark this answers is not in the document]
Comment (ggezer) [reply core/client/fs/src/main/java/alluxio/client/file/BaseFileSystem.java]: You can check the public constructor for OutStreamOptions, and there is now a Class.forName() call. I wrapped that code with try-catch and reverted this file to its original.

File: core/server/master/src/main/java/alluxio/master/file/options/CompleteFileContext.java (hunk 10 above)
Comment (ggezer) [reply core/server/master/src/main/java/alluxio/master/file/options/CompleteFileContext.java]: I put this on my task list for the consolidation work, as it'll cause lots of merge conflicts for pending PRs.

**Comment by ggezer** [comment]
@madanadit @aaudiber. Updated the PR. PTAL.

FYI, I postponed addressing some feedbacks to the a later PR in below schedule.
 1- FSMaster gRPC migration
 2- Rest of service migrations (BlockMaster/JobMaster/MetaMaster/KeyValueMaster)
 3- Security
 4- Consolidation (removing wire types/ operation context changes/ general improvement and cleaning up of code)

File: core/transport-options/src/grpc/file_system_master_options.proto (hunk 1 above)
Comment (ggezer) [reply core/transport-options/src/grpc/file_system_master_options.proto]: Looks like proto3 dropped support for checking presence of a value, so no more protoOption.hasQwe() family of methods. 
@madanadit , @aaudiber WDYT?
